runtime/v2/shim.go

```diff
@@ -205,31 +197,74 @@ type ShimInstance interface {
 	Client() any
 	// Delete will close the client and remove bundle from disk.
 	Delete(ctx context.Context) error
+	// Version returns shim's features compatibility version.
+	Version() int
 }
 
-func parseStartResponse(ctx context.Context, response []byte) (client.BootstrapParams, error) {
+func parseStartResponse(response []byte) (client.BootstrapParams, error) {
 	var params client.BootstrapParams
 
 	if err := json.Unmarshal(response, &params); err != nil || params.Version < 2 {
 		// Use TTRPC for legacy shims
 		params.Address = string(response)
 		params.Protocol = "ttrpc"
+		params.Version = 2
 	}
 
-	if params.Version > 2 {
+	if params.Version > CurrentShimVersion {
 		return client.BootstrapParams{}, fmt.Errorf("unsupported shim version (%d): %w", params.Version, errdefs.ErrNotImplemented)
 	}
 
 	return params, nil
 }
 
+// writeBootstrapParams writes shim's bootstrap configuration (e.g. how to connect, version, etc).
+func writeBootstrapParams(path string, params client.BootstrapParams) error {
+	path, err := filepath.Abs(path)
+	if err != nil {
+		return err
+	}
+	f, err := atomicfile.New(path, 0o666)
+	if err != nil {
+		return err
+	}
+
+	data, err := json.Marshal(&params)
+	if err != nil {
+		return err
+	}
```

Comment:
nit: reorder these so that a file isn't created if json marshalling fails.

Reply:
Addressed in https://github.com/containerd/containerd/pull/9233/commits/4db0537604b40992c8bb8de33d875d2e65122b4d